ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ProtobufQueryStreamResponseWriter.java

```diff
@@ -0,0 +1,166 @@
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.serde.connect.ConnectSchemas;
+import io.confluent.ksql.serde.connect.KsqlConnectSerializer;
+import io.confluent.ksql.serde.protobuf.ProtobufNoSRSerdeFactory;
+import io.confluent.ksql.util.KeyValue;
+import io.confluent.ksql.util.KeyValueMetadata;
+import io.vertx.core.http.HttpServerResponse;
+import java.util.List;
+import java.util.Optional;
+import org.apache.kafka.connect.data.ConnectSchema;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Struct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProtobufQueryStreamResponseWriter implements QueryStreamResponseWriter {
+
+  private static final Logger LOG
+          = LoggerFactory.getLogger(ProtobufQueryStreamResponseWriter.class);
+
+  private final HttpServerResponse response;
+  private final Optional<String> completionMessage;
+  private final Optional<String> limitMessage;
+  private ConnectSchema connectSchema;
+  private KsqlConnectSerializer<Struct> serializer;
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
+  public ProtobufQueryStreamResponseWriter(final HttpServerResponse response,
+                                           final Optional<String> completionMessage,
+                                           final Optional<String> limitMessage) {
+    this.response = response;
+    this.completionMessage = completionMessage;
+    this.limitMessage = limitMessage;
+  }
+
+  @Override
+  public QueryStreamResponseWriter writeMetadata(final QueryResponseMetadata metaData) {
+    final LogicalSchema schema = metaData.schema;
+    final String queryId = metaData.queryId;
+
+    connectSchema = ConnectSchemas.columnsToConnectSchema(schema.columns());
+    serializer = new ProtobufNoSRSerdeFactory(ImmutableMap.of())
+            .createSerializer(connectSchema, Struct.class, false);
+
+    final StreamedRow header = StreamedRow.headerProtobuf(
+            new QueryId(queryId), logicalToProtoSchema(schema));
+
+    response.write("[").write(serializeObject(header));
+    return this;
+  }
+
+  @Override
+  public QueryStreamResponseWriter writeRow(
+          final KeyValueMetadata<List<?>, GenericRow> row
+  ) {
+    final KeyValue<List<?>, GenericRow> keyValue = row.getKeyValue();
+    if (keyValue.value() == null) {
+      LOG.warn("Dropped tombstone. Not currently supported");
+    } else {
+      final Struct ksqlRecord = new Struct(connectSchema);
+      int i = 0;
+      for (Field field : connectSchema.fields()) {
+        ksqlRecord.put(
+                field,
+                keyValue.value().get(i));
+        i += 1;
+      }
+      final byte[] protoMessage = serializer.serialize("", ksqlRecord);
+      final Optional<DataRowProtobuf> protoRow =
+              StreamedRow.pullRowProtobuf(protoMessage).getRowProtobuf();
+      response.write(",\n").write(serializeObject(protoRow));
```

Comment:
is there any reason we're adding newlines to everything? the other formats seem to just rely on `,` in between rows and then users can pipe it into `jq` or something if they want to

I'm asking because it seems like we may be able to reuse a lot of the `JsonQueryStreamResponseWriter` code if we standardized

Reply:
Done. `JsonQueryStreamResponseWriter` now takes care of both json and protobuf format depending on the media type